File: Java/src/htmltest/base/CloneWalker_Base.java
package htmltest.base;

import static org.junit.Assert.fail;

import java.io.File;
import java.io.FileWriter;

import htmlwalker.exception.HtmlWalkerException;
import htmlwalker.exception.TagNotSupportedException;
import htmlwalker.platform.WalkerDocument;
import xhtmldomwalker.XhtmlDomPlatform;

public abstract class CloneWalker_Base extends CloneTest
{
	private final static String TSTPATH = "Test";

    public CloneWalker_Base() { super("html", "Base"); }
    public CloneWalker_Base(String ext) { super(ext, "Base"); }
    
    private String testHtml()
    {
        String eol = isXml() ? System.lineSeparator() : "\n";
        
    	String html = "";

        if (isXml())
    		html += "<html xmlns=\"http://www.w3.org/1999/xhtml\">" + eol;
        else
    		html += "<html>" + eol +" <head></head> " + eol;
        html += " <body> " + eol + "  <p>Hello, World!</p>  " + eol + " </body>" + eol + "</html>";
        
        return html;
    }

    @Override
    protected String testFolder() { return TSTPATH; }

    public void tUnsupported()
    {
        String html = "<html><body><unsupported></unsupported></body></html>";
        TestWalker walker = new TestWalker(platform());
        WalkerDocument document = platform().newDocument();

        try
        {
            document.loadHtml(html);
            walker.visit(document.documentTag());
            fail("Did not throw an exception");
        }
        catch (TagNotSupportedException e)
        {
            // Test passes
        }
        catch (HtmlWalkerException e)
        {
			e.printStackTrace();
			fail(e.getMessage());
		}
    }

    public void tWalk()
    {
        String html = testHtml();
        
        File testExpected = getTestOutput("expected.path_load." + this.ext);
        try
        {
            if (testExpected.exists())
            	testExpected.delete();

            try (FileWriter writer = new FileWriter(testExpected.getAbsolutePath()))
            {
                writer.write(html);
            }

            WalkerDocument document = platform().newDocument(testExpected.getPath());
            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());

            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;

            if (!expected.equals(walker.output().html()))
                fail("Expected - [" + expected + "] vs Out - [" + walker.output().html() + "]");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tLoadHtml()
    {
        String html = testHtml();

        try
        {
        	WalkerDocument document = platform().newDocument();
            document.loadHtml(html);
            
            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;
            
            if (!expected.equals(document.html()))
                fail("Expected - [" + expected + "] vs Out - [" + document.html() + "]");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tLoadPath()
    {
        String html = testHtml();
        
        File testExpected = getTestOutput("expected_load_path." + this.ext);
        try
        {
            if (testExpected.exists())
            	testExpected.delete();

            try (FileWriter writer = new FileWriter(testExpected.getAbsolutePath()))
            {
                writer.write(html);
            }

            WalkerDocument document = platform().newDocument();
            document.load(testExpected.getAbsolutePath());
                        
            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;
            
            if (!expected.equals(document.html()))
                fail("Expected - [" + expected + "] vs Out - [" + document.html() + "]");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tLoadHtmWalk()
    {
        String html = testHtml();
        
        try
        {
        	WalkerDocument document = platform().newDocument();
            document.loadHtml(html);

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());

            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;
            
            if (!expected.equals(document.html()))
                fail("Expected - [" + expected + "] vs Out - [" + document.html() + "]");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tLoadPathWalk()
    {
        String html = testHtml();

        File testExpected = getTestOutput("expected_load_path_walk." + this.ext);
        try
        {
            if (testExpected.exists())
            	testExpected.delete();

            try (FileWriter writer = new FileWriter(testExpected.getAbsolutePath()))
            {
                writer.write(html);
            }

            WalkerDocument document = platform().newDocument();
            document.load(testExpected.getAbsolutePath());

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());

            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;
            
            if (!expected.equals(document.html()))
                fail("Expected - [" + expected + "] vs Out - [" + document.html() + "]");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tSavePath()
    {
        String eol = System.lineSeparator();
        String html = "<html>" + eol;
        if (!isXml())
        	html += " <head></head> " + eol;
        html += " <body> " + eol;
        html += "  <p>Hello, World!</p>   " + eol; 
        html += " </body>" + eol;
        html += "</html>" + eol;

        try
        {
            File testOutput = getTestOutput("output_save_path." + this.ext);
            if (testOutput.exists())
                testOutput.delete();

            WalkerDocument document = platform().newDocument();
            document.loadHtml(html);

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());

            walker.output().save(testOutput.getAbsolutePath());

            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;
            
            compareToExpected(testOutput, expected);
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tSaveWriter()
    {
    	String eol = System.lineSeparator();
        String html = "<html>" + eol;
        if (!isXml())
        	html += " <head></head> " + eol;
        html += " <body> " + eol;
        html += "  <p>Hello, World!</p>   " + eol; 
        html += " </body>" + eol;
        html += "</html>" + eol;

        try
        {
            File testOutput = getTestOutput("output_save_writer." + this.ext);
            if (testOutput.exists())
                testOutput.delete();

            WalkerDocument document = platform().newDocument();
            document.loadHtml(html);

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());

            try (FileWriter writer = new FileWriter(testOutput.getAbsolutePath()))
            {
                walker.output().save(writer);
            }

            String expected = html;
            if (isXml())
            	expected = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator() + html;            

            compareToExpected(testOutput, expected);
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tEntityIsPreserved()
    {
        String html = "";
        if (isXml())
        	html = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator();
        
    	String eol = isXml() ? System.lineSeparator() : "\n";
    	html += "<html>" + eol;
    	if (!isXml())
    		html += " <head></head> " + eol;
        html += " <body> " + eol;
        html += "  <p>&lsquo;Hello, World!&rsquo;</p>  " + eol;
        html += " </body>" + eol;
        html += "</html>";

        try
        {
        	WalkerDocument document = platform().newDocument();
            document.loadHtml(html);

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());

            if (!html.equals(walker.output().html()))
                fail("Expected - [" + html + "] vs Out - [" + walker.output().html() + "]");
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tEntityIsPreservedSaved()
    {
        String html = "";
        if (isXml())
        	html = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator();
        
    	String eol = isXml() ? System.lineSeparator() : System.lineSeparator();
    	html += "<html>" + eol;
    	if (!isXml())
    		html += " <head></head> " + eol;
        html += " <body> " + eol;
        html += "  <p>&lsquo;Hello, World!&rsquo;</p>  " + eol;
        html += " </body>" + eol;
        html += "</html>" + eol;

        try
        {
            File testOutput = getTestOutput("output_entity." + this.ext);
            if (testOutput.exists())
                testOutput.delete();

            WalkerDocument document = platform().newDocument();
            document.loadHtml(html);

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());
            walker.output().save(testOutput.getAbsolutePath());

            compareToExpected(testOutput, html);
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    public void tEntityIsPreservedLoadSaved()
    {
        String html = "";
        if (isXml())
        	html = "<!DOCTYPE html PUBLIC \"" + XhtmlDomPlatform.DOCTYPE_PUBLIC + "\">" + System.lineSeparator();
        
    	String eol = isXml() ? System.lineSeparator() : "\n";
    	html += "<html>" + eol;
    	if (!isXml())
    		html += " <head></head> " + eol;
        html += " <body> " + eol;
        html += "  <p>&lsquo;Hello, World!&rsquo;</p>  " + eol;
        html += " </body>" + eol;
        html += "</html>";

        File testExpected = getTestOutput("expected_load_entity." + this.ext);
        try
        {
            if (testExpected.exists())
            	testExpected.delete();

            try (FileWriter writer = new FileWriter(testExpected.getAbsolutePath()))
            {
                writer.write(html);
            }

            File testOutput = getTestOutput("output_load_entity." + this.ext);
            if (testOutput.exists())
                testOutput.delete();

            WalkerDocument document = platform().newDocument();
            document.load(testExpected.getAbsolutePath());

            if (!isXml())
                html += System.lineSeparator();

            TestWalker walker = new TestWalker(platform());
            walker.visit(document.documentTag());
            walker.output().save(testOutput.getAbsolutePath());

            compareToExpected(testOutput, testExpected);
        }
        catch (Exception e)
        {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }
}

File: CSharp/HtmlTest/Base/CloneWalker_Base.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.IO;

namespace HtmlTest.Base
{
    public abstract class CloneWalker_Base : CloneTest
    {
        public CloneWalker_Base() : base("html", "Base") { }
        public CloneWalker_Base(string ext) : base(ext, "Base") { }


        public virtual void TUnsupported()
        {
            var html = @"<html><body><unsupported></unsupported></body></html>";
            var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
            var document = Platform.NewDocument();

            document.LoadHtml(html);

            try
            {
                walker.Visit(document.DocumentTag);
                Assert.Fail("Did not throw an exception");
            }
            catch (NotSupportedException)
            {
                // Test passes
            }
        }

        public virtual void TWalk()
        {
            var html = "<html><body><p>Hello, World!</p></body></html>";

            FileInfo testExpected = GetTestOutput("expected.path_load." + Ext);
            try
            {
                if (!testExpected.Exists)
                {
                    using var writer = new StreamWriter(testExpected.FullName);
                    writer.WriteLine(html);
                }

                var document = Platform.NewDocument(testExpected.FullName);
                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                if (walker.Output.Html.Replace(Environment.NewLine, "") != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + walker.Output.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TLoadHtml()
        {
            var html = "<html><body><p>Hello, World!</p></body></html>";

            try
            {
                var document = Platform.NewDocument();
                document.LoadHtml(html);

                if (document.Html != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + document.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TLoadPath()
        {
            var html = "<html><body><p>Hello, World!</p></body></html>";

            FileInfo testExpected = GetTestOutput("expected_load_path." + Ext);
            try
            {
                if (!testExpected.Exists)
                {
                    using var writer = new StreamWriter(testExpected.FullName);
                    writer.WriteLine(html);
                }

                var document = Platform.NewDocument();
                document.Load(testExpected.FullName);

                if (!Platform.IsXml)
                    html += Environment.NewLine;

                if (document.Html != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + document.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TLoadHtmWalk()
        {
            var html = "<html><body><p>Hello, World!</p></body></html>";

            try
            {
                var document = Platform.NewDocument();
                document.LoadHtml(html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                if (walker.Output.Html != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + walker.Output.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TLoadPathWalk()
        {
            var html = "<html><body><p>Hello, World!</p></body></html>";

            FileInfo testExpected = GetTestOutput("expected_load_path_walk." + Ext);
            try
            {
                if (!testExpected.Exists)
                {
                    using var writer = new StreamWriter(testExpected.FullName);
                    writer.WriteLine(html);
                }

                var document = Platform.NewDocument();
                document.Load(testExpected.FullName);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                if (walker.Output.Html.Replace(Environment.NewLine, "") != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + walker.Output.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TSavePath()
        {
            var html = "<html>  <body>    <p>Hello, World!</p>  </body></html>";

            try
            {
                FileInfo testOutput = GetTestOutput("output_save_path." + Ext);
                if (testOutput.Exists)
                    testOutput.Delete();

                var document = Platform.NewDocument();
                document.LoadHtml(html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                walker.Output.Save(testOutput.FullName);

                CompareToExpected(testOutput, html);
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TSaveTextWriter()
        {
            string html = "";
            if (Platform.IsXml)
                html += "<?xml version=\"1.0\" encoding=\"utf-8\"?>";
            html += "<html>  <body>    <p>Hello, World!</p>  </body></html>";

            try
            {
                FileInfo testOutput = GetTestOutput("output_save_writer." + Ext);
                if (testOutput.Exists)
                    testOutput.Delete();

                var document = Platform.NewDocument();
                document.LoadHtml(html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                using (var writer = new StreamWriter(testOutput.FullName))
                    walker.Output.Save(writer);

                CompareToExpected(testOutput, html);
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TEntityIsPreserved()
        {
            string hdr = "";
            if (Platform.IsXml)
                hdr = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
            var html = "<html><body><p>&lsquo;Hello, World!&rsquo;</p></body></html>";

            try
            {
                var document = Platform.NewDocument();
                document.LoadHtml(hdr + html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                if (walker.Output.Html != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + walker.Output.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TEntityIsPreservedSaved()
        {
            string hdr = "";
            if (Platform.IsXml)
                hdr = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
            var html = "<html>  <body>    <p>&lsquo;Hello, World!&rsquo;</p>  </body></html>";

            try
            {
                FileInfo testOutput = GetTestOutput("output_entity." + Ext);
                if (testOutput.Exists)
                    testOutput.Delete();

                var document = Platform.NewDocument();
                document.LoadHtml(hdr + html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);
                walker.Output.Save(testOutput.FullName);

                CompareToExpected(testOutput, html);
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TEntityIsPreservedLoadSaved()
        {
            string hdr = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
            var html = "<html><body><p>&lsquo;Hello, World!&rsquo;</p> </body></html>";

            var testExpected = GetTestOutput("expected_load_entity." + Ext);
            try
            {
                if (!testExpected.Exists)
                {
                    using var writer = new StreamWriter(testExpected.FullName);
                    if (Platform.IsXml)
                        writer.WriteLine(hdr);
                    writer.WriteLine(html);
                }

                var testOutput = GetTestOutput("output_load_entity." + Ext);
                if (testOutput.Exists)
                    testOutput.Delete();

                var document = Platform.NewDocument();
                document.Load(testExpected.FullName);

                if (!Platform.IsXml)
                    html += Environment.NewLine;

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);
                walker.Output.Save(testOutput.FullName);

                CompareToExpected(testOutput, testExpected);
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TNoEntityConversion()
        {
            string hdr = "";
            if (Platform.IsXml)
                hdr = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
            var html = "<html><body><p>“Hello, World!”</p></body></html>";

            try
            {
                var document = Platform.NewDocument();
                document.LoadHtml(hdr + html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                if (walker.Output.Html != html)
                    Assert.Fail("Expected - [" + html + "] vs Out - [" + walker.Output.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TNoEntityConversionSaved()
        {
            string hdr = "";
            if (Platform.IsXml)
                hdr = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
            var html = "<html>  <body>    <p>“Hello, World!”</p>  </body></html>";

            try
            {
                FileInfo testOutput = GetTestOutput("output_noentity." + Ext);
                if (testOutput.Exists)
                    testOutput.Delete();

                var document = Platform.NewDocument();
                document.LoadHtml(hdr + html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);
                walker.Output.Save(testOutput.FullName);

                CompareToExpected(testOutput, html);
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TNoEntityConversionLoadSaved()
        {
            string hdr = "";
            if (Platform.IsXml)
                hdr = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
            var html = "<html><body><p>“Hello, World!”</p></body></html>";

            FileInfo testExpected = GetTestOutput("expected_load_noentity." + Ext);
            try
            {
                if (!testExpected.Exists)
                {
                    using var writer = new StreamWriter(testExpected.FullName);
                    if (Platform.IsXml)
                        writer.WriteLine(hdr);
                    writer.WriteLine(html);
                }
                FileInfo testOutput = GetTestOutput("output_noentity." + Ext);
                if (testOutput.Exists)
                    testOutput.Delete();

                var document = Platform.NewDocument();
                document.LoadHtml(hdr + html);

                if (!Platform.IsXml)
                    html += Environment.NewLine;

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);
                walker.Output.Save(testOutput.FullName);

                CompareToExpected(testOutput, testExpected);
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        public virtual void TRemoveTag()
        {
            var html = "<html><body><p>Hello, World!</p><p>REMOVE THIS</p></body></html>";
            var expected = "<html><body><p>Hello, World!</p></body></html>";

            try
            {
                var document = Platform.NewDocument();
                document.LoadHtml(html);

                var walker = new CloneTestWalker(Platform, Platform.NewDocumentOptions());
                walker.Visit(document.DocumentTag);

                var body = walker.Output.DocumentTag.BodyTag;
                var removeThis = body.OwnedTags.ToList().Find(p => p.Name == "p" && p.OwnedText == "REMOVE THIS");
                body.OwnedTags.Remove(removeThis);

                if (walker.Output.Html != expected)
                    Assert.Fail("Expected - [" + expected + "] vs Out - [" + walker.Output.Html + "]");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }
    }
}
